core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -434,18 +440,47 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }.toMap
   }
 
-  def removeExpiredOffsets(startMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
-    val expiredOffsets = offsets
-      .filter {
+  def removeExpiredOffsets(currentTimestamp: Long, offsetRetentionMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
+
+    def getExpiredOffsets(baseTimestamp: CommitRecordMetadataAndOffset => Long): Map[TopicPartition, OffsetAndMetadata] = {
+      offsets.filter {
         case (topicPartition, commitRecordMetadataAndOffset) =>
-          commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp < startMs && !pendingOffsetCommits.contains(topicPartition)
-      }
-      .map {
+          !pendingOffsetCommits.contains(topicPartition) && {
+            commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
+              case None =>
+                // current version with no per partition retention
+                currentTimestamp - baseTimestamp(commitRecordMetadataAndOffset) >= offsetRetentionMs
```

Comment:
This is a little confusing. The function `baseTimestamp` can return -1 if no current state timestamp is set, which would cause this check to incorrectly succeed. This would be the case if we were loading the state of a group which existed before we began using the new group metadata schema. Are there any guards which prevent this case?

Basically the problem is that we need to handle the case that there is no expiration timestamp and there is no current state timestamp. I think in this case, `baseTimestamp` should just use the commit timestamp, which is basically like reverting to the old behavior. In other words, we should only be allowed to use the new behavior if `currentStateTimestamp` is defined.